src/app.controller.ts

```
@@ -1,13 +1,37 @@
-import { Controller, Get, Render } from '@nestjs/common';
+import { Body, Controller, Get, Post, Render } from '@nestjs/common';
 import { AppService } from './app.service';
+import { UserService } from './users/user.service';
+import { User } from './users/user.entity';
 
 @Controller()
 export class AppController {
-  constructor(private readonly appService: AppService) {}
+  constructor(
+    private readonly appService: AppService,
+    private readonly userService: UserService,
+  ) {}
 
   @Get()
   @Render('show')
   getHello(): object {
     return {};
   }
+
+  @Get('/register')
+  @Render('register')
+  registerGet(): object {
+    return {};
+  }
+
+  @Post('/register')
+  @Render('register')
+  async registerPost(@Body() body): Promise<object> {
+    const name = body['name'];
+    const email = body['email'];
+
+    const identifier = await this.appService.createAuth0User(email);
+
+    await this.userService.add(new User(email, name, identifier));
```

What happens if we create an Auth0 user, but the user creation fails? Or vice versa? It might be best to wrap it all in a transaction and:

* Part create the user with name and email
* Create the Auth0 user
* If the Auth0 call fails, rollback the transaction
* If the Auth0 call succeeds, add the identifier and save the User

---

We'll also probably want to raise an error and rescue from it to present an error back to the user if the API call fails, in a similar way to [how we handle validation errors](https://github.com/UKGovernmentBEIS/regulated-professions-register/blob/main/src/validation/validation-exception.filter.ts)

---

Where would user creation fail here? On the db write? In that case, I think it's always going to be possible to wind up with an Auth0 user being created, but no user being created in our DB

I do have in mind though that "email already exists in Auth0 but not in our db" is a case we'll need to handle

---

I think this is resolved in the latest push?